bokeh/util/structure.py

```diff
@@ -0,0 +1,315 @@
+            y_units="screen",
+            text_font_size="8pt",
+            text_font_style="italic",
+            text="Click on a model to see its attributes",
+        )
+        p2.add_layout(help)
+        p2.add_layout(edge_renderer)
+        p2.add_layout(node_renderer)
+        p2.tools.extend(
+            [node_hover_tool, tap_tool, BoxZoomTool(), ResetTool(), PanTool()]
+        )
+        p2.renderers.append(labels)
+        self._node_source = node_source
+        self._edge_source = edge_source
+        return p2
+
+    def _make_prop_dict(self):
+        """
+        Creates a dataframe of all the properties of all the
+        submodels of the model being
+        analyzed. Used as datasource to show attributes.
+        """
+        df = pd.DataFrame()
+        for x in self._graph.nodes(data=True):
+            M = self._model.select_one({"id": x[0]})
+            Z = pd.DataFrame(self.obj_props_to_df2(M))
+            Z["id"] = x[0]
+            Z["model"] = str(M)
+            Z["values"] = Z["values"].map(lambda x: str(x))
+            Z["types"] = Z["types"].map(lambda x: str(x))
+            df = df.append(Z)
+        return df
+
+    def _make_data_table(self):
+        """
+        Builds the datatable portion of the final plot.
+        """
+        columns = [
+            TableColumn(field="props", title="Property"),
+            TableColumn(field="values", title="Value"),
+        ]
+        prop_source = ColumnDataSource(self._prop_df)
+        model_id = self._node_source.data["index"][0]
+        groupfilter = GroupFilter(column_name="id", group=model_id)
+
+        data_table2_view = CDSView(source=prop_source, filters=[groupfilter])
+        data_table2 = DataTable(
+            source=prop_source,
+            view=data_table2_view,
+            columns=columns,
+            visible=False,
+            index_position=None,
+            fit_columns=True,
+            editable=False,
+        )
+
+        self._groupfilter = groupfilter
+        self._prop_source = prop_source
+        return data_table2
+
+    def _combined(self):
+        """
+        Connects the graph and the datatable with a simple CustomJS callback
+        so that clicking on a node/submodel narrows the view in the datatable to the
+        attributes associated with that submodel.
+        """
+
+        js_code = """const index = node_source.selected.indices[0];
+            f['group']=node_source.data['index'][index];
+            table['visible']=true;
+            prop_source.change.emit();"""
+
+        js = CustomJS(
+            args=dict(
+                node_source=self._node_source,
+                prop_source=self._prop_source,
+                f=self._groupfilter,
+                table=self._data_table,
+            ),
+            code=js_code,
+        )
+
+        self._node_source.selected.js_on_change("indices", js)
+        layout = column(self._graph_plot, self._data_table)
+        return layout
```

Comment:
This module needs at least minimal tests to exercise the public API (so that it can't change unintentionally)

Reply:
See also comment on the example about API

Reply:
@bryevdv I could use some advice on the right test.  I have a test that compares the submodel graph computed by the code against a known example, but that is based on the private API and it's more focused on the internals of the algorithm. The current setup has the public API returning a model.  I could precompute a structure graph for a simple model and have the test compare the result to this precomputed example -- is that what you have in mind?

Reply:
@jeremy9959  My comment was about the minmum bar, which is just:
> Any test that calls the function at all (so that if the function was removed or deleted unintentionally, that test would fail)
A easy next step would be to look at a return result and make sure it is the expected type (and might do so for several inputs to test corner cases)

More thorough would be what you describe, to actually compare return values. If that is something you can do, it'd definitely encourage that level of thoroughness.

Reply:
I put in a sort of compromise.  I compute the graph for a simple model and see if
it has the right number of submodels. See what you think.

Reply:
However, this reveals a new problem.  The networkx graph layout I am using needs pydot, and that doesn't seem to be available in the CI environment.

Reply:
And, similarly, pandas is shown as "module not found" in the tests.

Reply:
is the version of `pydot` you need available via conda? Then the `environment-test*.yml` files under the `ci` directory can be updated. If the test needs pandas then the pytest fixture `pd` should be added to the test arguments so that the test is skipped on the "minimal" test run (which specifically tests that basic functions work even if things like pandas are not installed)

For `isort` and general linter issues I def recommend running `pytest -m tests/codebase` before a push

Reply:
OK, I think I have it straightened out.